### DISCUSSION TITLE
Physical meaning of the solver options

### DISCUSSION THREAD

**Discussion by mwcoile** [reaktoro/reaktoro#421] [Q&A]
### Description

Hello,

I am new to Reaktoro. Thank you very much to Allan and those of you who contributed to such a nice equilibrium solver routine. It seems from my initial toying with it that the solver options "epsilon" and "logarithm_barrier_factor" are important to solver convergence in some cases, and so I am particularly interested in understanding them. 

_epsilon -- "The default lower bound for the amounts of the species"_
Is this the lower bound in moles for any species present at equilibrium? Or is it more complicated than that?

_logarithm_barrier_factor -- "The value multiplied by epsilon to compute the logarithm barrier penalty parameter tau"_
Increasing this value helps achieve solver convergence sometimes. I gradually increase it from the starting default value of 1.0 Does increasing this value impact the accuracy of the results? Intuitively, I expect that I might not want tau (the product of epsilon and this factor) to increase past 1.0, but I am unsure about what is going on under the hood. What is the practical upper bound at which I should stop increasing it?

Thank you in advance for any advice you can provide. 

### What is your operating system?

None

**Comment by allanleal** [comment]
Hi @mwcoile,  

The `epsilon` parameter sets a common lower bound for species amounts in moles. To customize lower and upper bounds for individual species, you can use an `EquilibriumRestrictions` object and pass it to the `EquilibriumSolver.solve` method. Just in case you find useful.  

The equilibrium algorithm in Reaktoro incorporates elements from logarithmic-barrier interior-point minimization algorithms. When `logarithm_barrier_factor=1`, the log-barrier parameter (`tau`) is equivalent to `epsilon`.  

For example:  
- If `epsilon = 1e-16` and `logarithm_barrier_factor=1`, then `tau = 1e-16`. Species with extremely small amounts (i.e., that should not exist in equilibrium or their mathematically correct amounts are extremely tiny, such as `1e-60`) will have values close to `1e-16`. Those that find themselves with amounts lower than `epsilon` are clipped to `epsilon` moles.  
- If `epsilon = 1e-16` and `logarithm_barrier_factor=100`, then `tau = 1e-14`. In this case, species with extremely small concentrations will have values closer to `1e-14`, and fewer species will fall below the `epsilon` threshold and be clipped (e.g., ending up with `1.23e-15` moles instead of 2.34e-17 moles).  

In some scenarios, reducing `epsilon` alone could achieve a similar effect to adjusting the `logarithm_barrier_factor`. 

The following paper could be helpful. The algorithm detailed in there, however, was used in Reaktoro v1, and not in v2. Similarities exist though. 

* _Leal, A. M. M., Kulik, D. A., Smith, W. R., & Saar, M. O. (2017). An overview of computational methods for chemical equilibrium and kinetic calculations for geochemical and reactive transport modeling. Pure and Applied Chemistry, 89(5), 597–643. https://doi.org/10.1515/pac-2016-1107_
